components/patch/default/index.mjs

```diff
@@ -1,43 +1,57 @@
-const _undefined = undefined;
 const { getPrototypeOf, defineProperty, getOwnPropertyDescriptor } = Reflect;
+
+const _undefined = undefined;
+
 export default (dependencies) => {
   const {
     expect: { expect },
     util: { hasOwnProperty },
   } = dependencies;
   return {
     patch: (object, key, value) => {
-      expect(
-        !hasOwnProperty(object, key),
-        "cannot patch object because it has %j as own property",
-        key,
-      );
-      let descriptor = _undefined;
-      for (
-        let prototype = object;
-        prototype !== null && descriptor === _undefined;
-        prototype = getPrototypeOf(prototype)
-      ) {
-        descriptor = getOwnPropertyDescriptor(prototype, key);
+      if (hasOwnProperty(object, key)) {
+        const descriptor = getOwnPropertyDescriptor(object, key);
+        expect(
+          hasOwnProperty(descriptor, "value"),
+          "cannot monkey-patch accessor property %j of object %o",
+          key,
+          object,
+        );
+        expect(
+          descriptor.configurable || descriptor.writable,
+          "cannot monkey-patch constant data property %j of object %o",
+          key,
+          object,
+        );
+        defineProperty(object, key, {
+          __proto__: descriptor,
+          value,
+        });
+        return descriptor.value;
+      } else {
+        defineProperty(object, key, {
+          __proto__: null,
```

Comment:
if the object to be monkey-patched already has the property, we reuse the same descriptor to avoid modifying too much the behavior of the object. If the object does not have the property we use a writable and configurable but not enumerable descriptor. In that case `__proto__: null` prevent the engine to go through `Object.prototype`.